cloudify_aws/ec2/keypair.py

```diff
@@ -0,0 +1,263 @@
+                        'but the key pair does not exist in the account: '
+                        '{0}'.format(str(e)))
+        else:
+            if key_file_in_filesystem:
+                raise NonRecoverableError(
+                        'Not external resource, '
+                        'but the key file exists locally.')
+            try:
+                self._get_key_pair_by_id(ctx.node.properties['resource_id'])
+            except NonRecoverableError:
+                pass
+            else:
+                raise NonRecoverableError(
+                        'Not external resource, '
+                        'but the key pair exists in the account.')
+
+        return True
+
+    def create(self, args=None, **_):
+        """Creates a keypair."""
+
+        key_pair_name = utils.get_resource_id()
+
+        kp = self.execute(self.client.create_key_pair,
+                          dict(key_name=key_pair_name), raise_on_falsy=True)
+
+        self._save_key_pair(kp)
+
+        ctx.instance.runtime_properties[constants.AWS_TYPE_PROPERTY] = \
+            constants.KEYPAIR['AWS_RESOURCE_TYPE']
+
+        self.resource_id = kp.name
+
+        return True
+
+    def created(self, args=None):
+
+        ctx.logger.info(
+                'Attempting to create keypair {0}.'
+                .format(self.cloudify_node_instance_id))
+
+        if self._create_external_keypair() or self.create(args):
+            return self.post_create()
+
+        raise NonRecoverableError(
+                'Neither external resource, nor Cloudify resource, '
+                'unable to create this resource.')
+
+    def delete(self, args=None, **_):
+        """Deletes a keypair."""
+
+        key_pair_name = utils.get_external_resource_id_or_raise(
+                'delete key pair', ctx.instance)
+
+        return self.execute(self.client.delete_key_pair,
+                            dict(key_name=key_pair_name), raise_on_falsy=True)
+
+    def post_delete(self):
+
+        super(KeyPair, self).post_delete()
+        self._delete_key_file()
+
+        return True
+
+    def _get_path_to_key_file(self):
+        """Gets the path to the key file.
+
+        :param ctx: The Cloudify context.
+        :returns key_path: Path to the key file.
+        :raises NonRecoverableError: If private_key_path is not set.
+        """
+
+        if 'private_key_path' not in ctx.node.properties:
+            raise NonRecoverableError(
+                    'Unable to get key file path, private_key_path not set.')
+
+        return os.path.expanduser(ctx.node.properties['private_key_path'])
+
+    def _search_for_key_file(self, path_to_key_file):
+        """ Checks if the key_path exists in the local filesystem.
+
+        :param key_path: The path to the key pair file.
+        :return boolean if key_path exists (True) or not.
+        """
+
+        return True if os.path.exists(path_to_key_file) else False
+
+    def _get_key_pair_by_id(self, key_pair_id):
```

Comment:
you don't want this function, you want to use the built-in function get_all_matching